nuxeo-features/nuxeo-platform-oauth/src/main/java/org/nuxeo/ecm/platform/oauth2/tokens/OAuth2TokenServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * (C) Copyright 2019 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Salem Aouana
+ */
+
+package org.nuxeo.ecm.platform.oauth2.tokens;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.query.sql.model.Predicates;
+import org.nuxeo.ecm.core.query.sql.model.QueryBuilder;
+import org.nuxeo.ecm.directory.Session;
+import org.nuxeo.ecm.directory.api.DirectoryService;
+import org.nuxeo.ecm.platform.oauth2.enums.NuxeoOAuth2TokenType;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+/**
+ * The implementation that manages the tokens in Nuxeo.
+ *
+ * @since 11.1
+ */
+public class OAuth2TokenServiceImpl extends DefaultComponent implements OAuth2TokenService {
+
+    @Override
+    public List<NuxeoOAuth2Token> getTokens() {
+        return findTokens(null, null);
+    }
+
+    @Override
+    public List<NuxeoOAuth2Token> getTokens(String nxuser) {
+        requireNonNull(nxuser, "nxuser cannot be null");
+
+        return findTokens(nxuser, null);
+    }
+
+    @Override
+    public List<NuxeoOAuth2Token> getTokens(NuxeoOAuth2TokenType type) {
+        requireNonNull(type, "oAuth2TokenType cannot be null");
+
+        return findTokens(null, type);
+    }
+
+    @Override
+    public List<NuxeoOAuth2Token> getTokens(String nxuser, NuxeoOAuth2TokenType type) {
+        requireNonNull(nxuser, "nxuser cannot be null");
+        requireNonNull(type, "oAuth2TokenType cannot be null");
+
+        return findTokens(nxuser, type);
+    }
+
+    protected List<NuxeoOAuth2Token> findTokens(String nxuser, NuxeoOAuth2TokenType type) {
```

Comment:
The OAuth2Object#checkPermission logic should move to the service impl... Plus add a comment to document the use of Framework.doPrivileged here

Reply:
> The OAuth2Object#checkPermission logic should move to the service impl...

Not sure:
- The user is not always available in the service's methods.
- For consistency, all the `OAuth2Object` methods are calling it directly.

Reply:
After discussing it with @efge and @guirenard, it seems indeed a good practice to have the permission check in the service, because:
- It prevents security holes, in particular from potential callers that do not really know what they do, e.g. from some JS code.
- It eases security audit to have the check as close as possible to the privileged code.
- It makes it clearer that a principal is needed in the context of the method.

@RSalem07 we can see how to rework it together.